### DISCUSSION TITLE
How can i  keep npm packages with tailwind.css working

### DISCUSSION THREAD

**Discussion by hellomrbigshot** [tailwindlabs/tailwindcss#8814] [Help] [answered]
I creat a privite npm package(PackageA) with tailwind. There are some media query classes like ```hidden lg:flex```.
Then I create another project ProjectB with tailwind, then I import PackageA in this project in ```main.ts```.

```
import { createApp } from 'vue'
import App from './App.vue'
import PackageA from 'PackageA' // import component
import ''./index.css' // import tailwind.css
import 'PackageA/dist/style.css' // import PackageA css

...

createApp(App).use(PackageA).mount('#app')
```

I use ```hidden``` in ProjectB but no ```lg:flex```

After I build ProjectB, I find ```lg:flex``` doesn't work. Also ```PackageA/dist/style.css``` has been bundled, but ```hidden``` in ProjectB covers ```hidden lg:flex```
I can add 'lg:flex' in ```tailwind.config.js``` to solve this problem, but i think it's not a good solution.
```
purge: {
  safelist: ['lg:flex']
}
```

I Try to add 'PackageA/dist/style.css' in ```purge.content``` but it doesn't work.
```
purge: {
    content: ['./public/**/*.html', './src/**/*.{vue,js,ts,jsx,tsx}',  './node_modules/PackageA/dist/style.css']
}
```
Is there anyway to solve this problem.

**Comment by ghost** [comment, marked as answer]
I have found the solution👇

https://github.com/tailwindlabs/tailwindcss/discussions/4389

my tailwind.config.js look likes below:
```diff
module.exports = {
  purge: {
    enable: process.env.NODE_ENV !== 'development',
    content: [
      './src/**/*.html',
      './src/**/*.vue',
+     './node_modules/my-component-path/**/*.{vue,js,ts,jsx,tsx}',
    ],
  },
  // ...
}
```

**Reply by copostic** [reply]
Shouldn't we use a method like this one ?
https://dev.to/samrobbins85/making-an-npm-package-for-a-react-component-library-with-tailwind-css-a0h

**Reply by AurelianSpodarec** [reply]
What about with TailwindCSS V4?

**Reply by hellomrbigshot** [reply]
Yeah! It works for me😄😄😄

**Reply by cjcheshire** [reply]
Will this possibly end up adding css rules for components possible not used in the app?

**Reply by Queen-esther01** [reply]
> What about with TailwindCSS V4?

Heyy, this issue has bugged me for over 1 month. Allow me to proudly present the solution to your problem : )

As you know, V4 doesnt make use of the config file anymore, so what you need to do is add this line `@source "../../node_modules/<package-name>"` to your globals.css file or whatever file holds your `@import "tailwindcss"` command.

Add this command and watch your style magically fall in place lol. I really hope this works for you too. Here's the supporting documentation: https://tailwindcss.com/docs/detecting-classes-in-source-files#explicitly-registering-sources

**Comment by drusellers** [comment]
Having to map my explicit node_modules path seems a bit awkward. Is there a way to use the existing path resolution features that we use with other packages?

```
module.exports = {
  purge: {
    content: [
      './src/**/*.html',
      './src/**/*.vue',
      'my-component/**/*.{vue,js,ts,jsx,tsx}', // would pull in the npm package `my-component`
      '@scope/my-component/**/*.{vue,js,ts,jsx,tsx}', // would pull in the npm package `@scope/my-component`
    ],
  },
  // ...
}
```

**Reply by rojvv** [reply]
Node.js seems a bit awkward.

**Reply by freshgiammi** [reply]
Have you looked at this? I've had success using the path module to resolve a package inside a monorepo. https://tailwindcss.com/docs/content-configuration#working-with-third-party-libraries

My config looks like this (mind you I'm importing components from the src folder and not the dist otherwise I don't get docgen, but you could consume from dist by removing the `/src` suffix and looking for `.js` files instead:
```cjs
const path = require("path");

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [
    "./src/stories/**/*.{ts,tsx,mdx}",
    path.join(path.dirname(require.resolve("@acme/core/src")), "**/*.{ts,tsx}"),
  ],
  theme: {
    extend: {},
  },
  plugins: [],
};
```

**Reply by rojvv** [reply]
What about ESM?

**Reply by freshgiammi** [reply]
`path.join(path.dirname(require.resolve("@acme/core")), "**/*.{mjs}")` works fine for me loading modules from the output folder, but mind you I've only tested this in the monorepo I'm working on right now.

**Reply by drusellers** [reply]
That is def closer. hmm. TY

**Reply by GeorgeTaveras1231** [reply]
Just calling out that none of the approaches shared here work well with package exports or PNP. 

Because `require.resolve('my-package')` may resolve to any arbitrary file inside the package, and the resolution will use the condition used by require (the `require` condition in `package.json['exports']`) which may not be the same condition used by a bundler (most bundlers will use the `module` condition before falling back to the `require` condition). This is a problem if a package is structured as 
```
<root>
|_ cjs
  |_ component.cjs
|_ module
  |_ component.js // (or component.mjs)
```

In addition, attempting to resolve a specific folder inside a package (like `require.resolve('@acme/core/src')` or `require.resolve('package/module')` requires that the author explicitly provide this export path. 

Lastly, writing `./node_modules/package/**/*.js` doesn't play well with PnP which does not store modules in the `node_modules` folder. 

Ideally, tailwind would support package resolution natively, so that it can handle resolving the files properly. And maybe even support a custom `tailwind` package export condition to allow package authors to explicitly list the files that can be processed by tailwind 💡 

Package authors would write something like

// package.json
```json
{
  "name": "@acme/core",
  "exports": {
    "./**/*.js": {
      "tailwind": "./path/to/my-es-component-files/**/*.js"
    }
  }
}
```

And consumers of the package would write

// tailwind.config.js
```js
module.exports = {
  content: [
    "@acme/core"
  ]
}
```

**Reply by AurelianSpodarec** [reply]
Do you have a solution with V4?

**Comment by LouisLanganay** [comment]
It looks like the issue you're encountering is due to conflicting Tailwind CSS classes between your private npm package (PackageA) and your ProjectB setup. When importing PackageA into ProjectB, the styles from PackageA's `style.css` are overriding certain Tailwind CSS utilities like `lg:flex`, causing unexpected behavior.

Here are two possible solutions to address this problem:

# Solution 1: Use Tailwind's `prefix` option

You can configure Tailwind CSS to prefix all classes generated by Tailwind with a specific prefix. This ensures that classes from different sources (like PackageA and ProjectB) don't conflict with each other.

## 1. Modify your Tailwind CSS configuration file (`tailwind.config.js`):

   ```javascript
   // tailwind.config.js
   module.exports = {
     // other configurations
     prefix: 'tw-', // Replace 'tw-' with your desired prefix
   };
   ```

   With this setup, every Tailwind CSS class in your application will be prefixed with `tw-`, ensuring uniqueness.

### Example usage:

   ```html
   <!-- Before -->
   <div class="hidden lg:flex gap-5" />

   <!-- After applying prefix -->
   <div class="tw-hidden lg:flex tw-gap-5" />
   ```

# Solution 2: Rename classes during PostCSS compilation

Another approach is to rename classes during the PostCSS compilation step. This can be achieved using a custom plugin in your `postcss.config.js` file:

## 1. Update your `postcss.config.js` to include a renaming plugin:

   ```javascript
   // postcss.config.js
   module.exports = {
     plugins: [
       require('tailwindcss'),
       require('autoprefixer'),
       function (css) {
         css.walkRules((rule) => {
           if (rule.selector.startsWith('.')) {
             rule.selector = rule.selector
               .split(' ')
               .map(selector => {
                 if (selector.startsWith('.')) {
                   const prefix = 'obf-'; // Replace 'obf-' with your desired prefix
                   return '.' + prefix + selector.substr(1);
                 }
                 return selector;
               })
               .join(' ');
           }
         });
       }
     ]
   };
   ```

   This plugin renames all CSS classes with a specified prefix ('obf-' in this example), preventing conflicts between PackageA and ProjectB styles.

   ## 2.Utilize a function to apply this renaming to your components:

   ```typescript
   import clsx, { ClassValue } from "clsx";
   import { twMerge } from "tailwind-merge";

   const obfuscateClassName = (className: string, prefix: string) => `${prefix}${className}`;

   export function cn(...inputs: ClassValue[]) {
     const prefix = 'obf-'; // Replace 'obf-' with your desired prefix
     const classNames = twMerge(clsx(inputs));
     return classNames.split(' ').map((className) => obfuscateClassName(className, prefix)).join(' ');
   }
   ```

   ### Exemple:

   ```html
   <!-- Before -->
   <div class="h-5 w-5 inline-block mr-2" />

   <!-- After applying prefix -->
   <div class="obf-h-5 obf-w-5 obf-inline-block obf-mr-2" />
   ```

Using either of these solutions should help resolve the issue of conflicting Tailwind CSS classes between PackageA and ProjectB, ensuring that styles are applied correctly without overriding each other.

Let me know if you have any questions or need further assistance!

**Reply by craigrileyuk** [reply]
This is nothing to do with the question. Don't post Chat-GPT spam, especially when it's not remotely relevant.

**Reply by AurelianSpodarec** [reply]
Especially on GitHub, this guy trying to farm fake internet emojs lol got got a thumbs up, in reverse : d

Anyone has a solution for V4?

**Comment by nikolas-beckel** [comment]
Hey, is there a recommended way by now or do I still have to refer to my node_modules folder? Thank you.

**Reply by craigrileyuk** [reply]
Best suggestions I've seen so far involve either (1) Getting the user to wrap their entire `export default {}` object in a function and you modifying the config before it's passed to Tailwind, or (2) Give them a small exported function from your package that returns the necessary content string, e.g.

```js
import { tailwindContent } from "@yourname/yourpackage";

export default {
    content: [
        "./resources/js/**/*.{js,vue}",
        tailwindContent() // will return something like "./node_modules/yourname/yourpackage/dist/**/*.{js}"
    ]
}
```

Personally I prefer the second approach since it doesn't mess with their existing config and you can still update the content string if you like. If you envision your project needing multiple content strings, you can always use spreading from the off:

```js
    content: [
        ...tailwindContent()
    ]
```

and have your function return an array of strings.
